Crash when returning if date picker was left open

How to reproduce:
1. Create/Edit a card
2. Go to the Options tab
3. Click "Choose expiry date" under "Expiry date"
4. Switch to another app
5. Switch back to Catima

Device:
Fairphone 3, LineageOS for microG 19.1 (Android 12.1)

Crash:
```
FATAL EXCEPTION: main
Process: me.hackerchick.catima, PID: 29237
java.lang.RuntimeException: Unable to start activity ComponentInfo{me.hackerchick.catima/protect.card_locker.LoyaltyCardEditActivity}: androidx.fragment.app.Fragment$InstantiationException: Unable to instantiate fragment protect.card_locker.LoyaltyCardEditActivity$DatePickerFragment: could not find Fragment constructor
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3707)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3864)
	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2253)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7870)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
Caused by: androidx.fragment.app.Fragment$InstantiationException: Unable to instantiate fragment protect.card_locker.LoyaltyCardEditActivity$DatePickerFragment: could not find Fragment constructor
	at androidx.fragment.app.Fragment.instantiate(Fragment.java:628)
	at androidx.fragment.app.FragmentContainer.instantiate(FragmentContainer.java:57)
	at androidx.fragment.app.FragmentManager$3.instantiate(FragmentManager.java:483)
	at androidx.fragment.app.FragmentStateManager.<init>(FragmentStateManager.java:85)
	at androidx.fragment.app.FragmentManager.restoreSaveState(FragmentManager.java:2728)
	at androidx.fragment.app.FragmentController.restoreSaveState(FragmentController.java:198)
	at androidx.fragment.app.FragmentActivity$2.onContextAvailable(FragmentActivity.java:149)
	at androidx.activity.contextaware.ContextAwareHelper.dispatchOnContextAvailable(ContextAwareHelper.java:99)
	at androidx.activity.ComponentActivity.onCreate(ComponentActivity.java:322)
	at androidx.fragment.app.FragmentActivity.onCreate(FragmentActivity.java:273)
	at protect.card_locker.CatimaAppCompatActivity.onCreate(CatimaAppCompatActivity.java:24)
	at protect.card_locker.LoyaltyCardEditActivity.onCreate(LoyaltyCardEditActivity.java:308)
	at android.app.Activity.performCreate(Activity.java:8057)
	at android.app.Activity.performCreate(Activity.java:8037)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1341)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3688)
	... 12 more
Caused by: java.lang.NoSuchMethodException: protect.card_locker.LoyaltyCardEditActivity$DatePickerFragment.<init> []
	at java.lang.Class.getConstructor0(Class.java:2363)
	at java.lang.Class.getConstructor(Class.java:1759)
	at androidx.fragment.app.Fragment.instantiate(Fragment.java:613)
	... 27 more
```

--- a/app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java
+++ b/app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java
@@ -1352,11 +1352,15 @@ public class LoyaltyCardEditActivity extends CatimaAppCompatActivity implements 
     public static class DatePickerFragment extends DialogFragment
             implements DatePickerDialog.OnDateSetListener {
 
+        public interface OnDatePickListener {
+            void onDatePicked(@NonNull LoyaltyCardField textFieldToEdit, @NonNull Date newDate);
+        }
+
         private static final String TEXT_FIELD_TO_EDIT_ARGUMENT_KEY = "text_field_to_edit";
         private static final String CURRENT_DATE_ARGUMENT_KEY = "current_date";
         private static final String MIN_DATE_ARGUMENT_KEY = "min_date";
         private static final String MAX_DATE_ARGUMENT_KEY = "max_date";
-        private static final String PICK_DATE_REQUEST_KEY = "pick_date_request_key";
+        private static final String PICK_DATE_REQUEST_KEY = "pick_date_request";
         private static final String NEWLY_PICKED_DATE_ARGUMENT_KEY = "newly_picked_date";
 
         LoyaltyCardField textFieldEdit;
@@ -1374,10 +1378,6 @@ public class LoyaltyCardEditActivity extends CatimaAppCompatActivity implements 
             DatePickerFragment fragment = new DatePickerFragment();
             fragment.setArguments(args);
             return fragment;
-        }
-
-        public interface OnDatePickListener {
-            void onDatePicked(@NonNull LoyaltyCardField textFieldToEdit, @NonNull Date newDate);
         }
 
         public static void registerDatePickListener(@NonNull AppCompatActivity activity, @NonNull OnDatePickListener listener) {

--- a/app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java
+++ b/app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java
@@ -36,6 +36,20 @@ import android.widget.ImageView;
 import android.widget.TextView;
 import android.widget.Toast;
 
+import androidx.activity.result.ActivityResultLauncher;
+import androidx.activity.result.contract.ActivityResultContracts;
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.annotation.StringRes;
+import androidx.appcompat.app.AlertDialog;
+import androidx.appcompat.app.AppCompatActivity;
+import androidx.appcompat.widget.AppCompatTextView;
+import androidx.appcompat.widget.Toolbar;
+import androidx.core.content.ContextCompat;
+import androidx.core.content.FileProvider;
+import androidx.exifinterface.media.ExifInterface;
+import androidx.fragment.app.DialogFragment;
+
 import com.google.android.material.chip.Chip;
 import com.google.android.material.chip.ChipGroup;
 import com.google.android.material.color.MaterialColors;
@@ -67,20 +81,9 @@ import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Locale;
 import java.util.NoSuchElementException;
+import java.util.Objects;
 import java.util.concurrent.Callable;
 
-import androidx.activity.result.ActivityResultLauncher;
-import androidx.activity.result.contract.ActivityResultContracts;
-import androidx.annotation.NonNull;
-import androidx.annotation.Nullable;
-import androidx.annotation.StringRes;
-import androidx.appcompat.app.AlertDialog;
-import androidx.appcompat.widget.AppCompatTextView;
-import androidx.appcompat.widget.Toolbar;
-import androidx.core.content.ContextCompat;
-import androidx.core.content.FileProvider;
-import androidx.exifinterface.media.ExifInterface;
-import androidx.fragment.app.DialogFragment;
 import protect.card_locker.async.TaskHandler;
 import protect.card_locker.databinding.LayoutChipChoiceBinding;
 import protect.card_locker.databinding.LoyaltyCardEditActivityBinding;
@@ -375,6 +378,17 @@ public class LoyaltyCardEditActivity extends CatimaAppCompatActivity implements 
         addDateFieldTextChangedListener(validFromField, R.string.anyDate, R.string.chooseValidFromDate, LoyaltyCardField.validFrom);
 
         addDateFieldTextChangedListener(expiryField, R.string.never, R.string.chooseExpiryDate, LoyaltyCardField.expiry);
+
+        DatePickerFragment.registerDatePickListener(this, (textFieldToEdit, newDate) -> {
+            switch (textFieldToEdit) {
+                case validFrom:
+                    formatDateField(this, validFromField, newDate);
+                    break;
+                case expiry:
+                    formatDateField(this, expiryField, newDate);
+                    break;
+            }
+        });
 
         balanceField.setOnFocusChangeListener((v, hasFocus) -> {
             if (!hasFocus) {
@@ -967,10 +981,9 @@ public class LoyaltyCardEditActivity extends CatimaAppCompatActivity implements 
                     if (!lastValue.toString().equals(getString(chooseDateOptionStringId))) {
                         dateField.setText(lastValue);
                     }
-                    DialogFragment datePickerFragment = new DatePickerFragment(
-                            LoyaltyCardEditActivity.this,
-                            dateField,
-                            // if the expiry date is being set, set date picker's minDate to the 'valid from' date
+                    DialogFragment datePickerFragment = DatePickerFragment.newInstance(
+                            loyaltyCardField,
+                            (Date) dateField.getTag(),
                             loyaltyCardField == LoyaltyCardField.expiry ? (Date) validFromField.getTag() : null,
                             // if the 'valid from' date is being set, set date picker's maxDate to the expiry date
                             loyaltyCardField == LoyaltyCardField.validFrom ? (Date) expiryField.getTag() : null);
@@ -1336,27 +1349,54 @@ public class LoyaltyCardEditActivity extends CatimaAppCompatActivity implements 
     public static class DatePickerFragment extends DialogFragment
             implements DatePickerDialog.OnDateSetListener {
 
-        final Context context;
-        final EditText textFieldEdit;
-        @Nullable
-        final Date minDate;
-        @Nullable
-        final Date maxDate;
+        private static final String TEXT_FIELD_TO_EDIT_ARGUMENT_KEY = "text_field_to_edit";
+        private static final String CURRENT_DATE_ARGUMENT_KEY = "current_date";
+        private static final String MIN_DATE_ARGUMENT_KEY = "min_date";
+        private static final String MAX_DATE_ARGUMENT_KEY = "max_date";
+        private static final String PICK_DATE_REQUEST_KEY = "pick_date_request_key";
+        private static final String NEWLY_PICKED_DATE_ARGUMENT_KEY = "newly_picked_date";
 
-        DatePickerFragment(Context context, EditText textFieldEdit, @Nullable Date minDate, @Nullable Date maxDate) {
-            this.context = context;
-            this.textFieldEdit = textFieldEdit;
-            this.minDate = minDate;
-            this.maxDate = maxDate;
+        public static DatePickerFragment newInstance(@NonNull LoyaltyCardField textField, @Nullable Date currentDate, @Nullable Date minDate, @Nullable Date maxDate) {
+            Bundle args = new Bundle();
+            args.putSerializable(TEXT_FIELD_TO_EDIT_ARGUMENT_KEY, textField);
+            args.putSerializable(CURRENT_DATE_ARGUMENT_KEY, currentDate);
+            args.putSerializable(MIN_DATE_ARGUMENT_KEY, minDate);
+            args.putSerializable(MAX_DATE_ARGUMENT_KEY, maxDate);
+            DatePickerFragment fragment = new DatePickerFragment();
+            fragment.setArguments(args);
+            return fragment;
         }
+
+        public interface OnDatePickListener {
+            void onDatePicked(@NonNull LoyaltyCardField textFieldToEdit, @NonNull Date newDate);
+        }
+
+        public static void registerDatePickListener(@NonNull AppCompatActivity activity, @NonNull OnDatePickListener listener) {
+            activity.getSupportFragmentManager().setFragmentResultListener(
+                    PICK_DATE_REQUEST_KEY,
+                    activity,
+                    (requestKey, result) -> listener.onDatePicked(
+                            (LoyaltyCardField) Objects.requireNonNull(result.getSerializable(TEXT_FIELD_TO_EDIT_ARGUMENT_KEY)),
+                            (Date) Objects.requireNonNull(result.getSerializable(NEWLY_PICKED_DATE_ARGUMENT_KEY))));
+        }
+
+        LoyaltyCardField textFieldEdit;
+        @Nullable
+        Date minDate;
+        @Nullable
+        Date maxDate;
 
         @NonNull
         @Override
         public Dialog onCreateDialog(Bundle savedInstanceState) {
+            Bundle args = requireArguments();
+            textFieldEdit = (LoyaltyCardField) args.getSerializable(TEXT_FIELD_TO_EDIT_ARGUMENT_KEY);
+            minDate = (Date) args.getSerializable(MIN_DATE_ARGUMENT_KEY);
+            maxDate = (Date) args.getSerializable(MAX_DATE_ARGUMENT_KEY);
             // Use the current date as the default date in the picker
             final Calendar c = Calendar.getInstance();
 
-            Date date = (Date) textFieldEdit.getTag();
+            Date date = (Date) args.getSerializable(CURRENT_DATE_ARGUMENT_KEY);
             if (date != null) {
                 c.setTime(date);
             }
@@ -1398,7 +1438,10 @@ public class LoyaltyCardEditActivity extends CatimaAppCompatActivity implements 
 
             Date date = new Date(unixTime);
 
-            formatDateField(context, textFieldEdit, date);
+            Bundle result = new Bundle();
+            result.putSerializable(TEXT_FIELD_TO_EDIT_ARGUMENT_KEY, textFieldEdit);
+            result.putSerializable(NEWLY_PICKED_DATE_ARGUMENT_KEY, date);
+            getParentFragmentManager().setFragmentResult(PICK_DATE_REQUEST_KEY, result);
         }
     }
 

--- a/app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java
+++ b/app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java
@@ -387,6 +387,8 @@ public class LoyaltyCardEditActivity extends CatimaAppCompatActivity implements 
                 case expiry:
                     formatDateField(this, expiryField, newDate);
                     break;
+                default:
+                    throw new AssertionError("Unexpected field: " + textFieldToEdit);
             }
         });
 

--- a/app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java
+++ b/app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java
@@ -986,6 +986,7 @@ public class LoyaltyCardEditActivity extends CatimaAppCompatActivity implements 
                     DialogFragment datePickerFragment = DatePickerFragment.newInstance(
                             loyaltyCardField,
                             (Date) dateField.getTag(),
+                            // if the expiry date is being set, set date picker's minDate to the 'valid from' date
                             loyaltyCardField == LoyaltyCardField.expiry ? (Date) validFromField.getTag() : null,
                             // if the 'valid from' date is being set, set date picker's maxDate to the expiry date
                             loyaltyCardField == LoyaltyCardField.validFrom ? (Date) expiryField.getTag() : null);
@@ -1358,6 +1359,12 @@ public class LoyaltyCardEditActivity extends CatimaAppCompatActivity implements 
         private static final String PICK_DATE_REQUEST_KEY = "pick_date_request_key";
         private static final String NEWLY_PICKED_DATE_ARGUMENT_KEY = "newly_picked_date";
 
+        LoyaltyCardField textFieldEdit;
+        @Nullable
+        Date minDate;
+        @Nullable
+        Date maxDate;
+
         public static DatePickerFragment newInstance(@NonNull LoyaltyCardField textField, @Nullable Date currentDate, @Nullable Date minDate, @Nullable Date maxDate) {
             Bundle args = new Bundle();
             args.putSerializable(TEXT_FIELD_TO_EDIT_ARGUMENT_KEY, textField);
@@ -1381,12 +1388,6 @@ public class LoyaltyCardEditActivity extends CatimaAppCompatActivity implements 
                             (LoyaltyCardField) Objects.requireNonNull(result.getSerializable(TEXT_FIELD_TO_EDIT_ARGUMENT_KEY)),
                             (Date) Objects.requireNonNull(result.getSerializable(NEWLY_PICKED_DATE_ARGUMENT_KEY))));
         }
-
-        LoyaltyCardField textFieldEdit;
-        @Nullable
-        Date minDate;
-        @Nullable
-        Date maxDate;
 
         @NonNull
         @Override

